fastcrypto-tbls/src/weight_reduction/mod.rs

```diff
@@ -0,0 +1,539 @@
+            self.tickets.push(0);
+        }
+        self.tickets[index] += 1;
+        self.total += 1;
+    }
+
+    fn get(&self, index: usize) -> u64 {
+        match self.tickets.get(index) {
+            Some(&x) => x,
+            None => 0,
+        }
+    }
+
+    fn data(&self) -> &[u64] {
+        &self.tickets
+    }
+
+    fn extract_data(self) -> Vec<u64> {
+        self.tickets
+    }
+
+    fn total(&self) -> u64 {
+        self.total
+    }
+
+    #[cfg(test)]
+    fn from_vec(tickets: Vec<u64>) -> Self {
+        let total = tickets.iter().sum();
+        Self { tickets, total }
+    }
+}
+
+// Helper types for generating deltas
+#[derive(Eq, PartialEq)]
+struct QueueElement {
+    s: Ratio,
+    i: usize,
+}
+
+impl Ord for QueueElement {
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.s.cmp(&self.s).then(other.i.cmp(&self.i))
+    }
+}
+
+impl PartialOrd for QueueElement {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+struct Generator<'a> {
+    weights: &'a [u64],
+    c: Ratio,
+    r: usize,
+    queue: BinaryHeap<QueueElement>,
+}
+
+impl<'a> Generator<'a> {
+    fn new(weights: &'a [u64], c: Ratio) -> Self {
+        assert!(!weights.is_empty());
+        debug_assert!(weights.windows(2).all(|w| w[0] >= w[1]));
+        assert!(*weights.last().unwrap() > 0);
+        assert!(c >= 0.into());
+        assert!(c < 1.into());
+
+        let mut queue = BinaryHeap::new();
+        queue.push(QueueElement {
+            s: (Ratio::from_integer(1) - c) / Ratio::from_integer(*weights.first().unwrap()),
+            i: 0,
+        });
+
+        Self {
+            weights,
+            c,
+            r: 0,
+            queue,
+        }
+    }
+}
+
+impl Iterator for Generator<'_> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let elem = self.queue.pop().unwrap();
+        let new_value = (elem.s * self.weights[elem.i] + self.c).to_integer();
+
+        self.queue.push(QueueElement {
+            s: (Ratio::from_integer(new_value + 1) - self.c)
+                / Ratio::from_integer(self.weights[elem.i]),
+            i: elem.i,
+        });
+        if (elem.i == self.r) && (self.r + 1 < self.weights.len()) {
+            self.r += 1;
+            self.queue.push(QueueElement {
+                s: (Ratio::from_integer(1) - self.c) / Ratio::from_integer(self.weights[self.r]),
+                i: self.r,
+            });
+        }
+
+        Some(elem.i)
+    }
+}
+
+fn generate_deltas(weights: &[u64], c: Ratio) -> impl Iterator<Item = usize> + '_ {
+    Generator::new(weights, c)
+}
+
+// Calculates the head indices for the current batch.
+fn calc_indices_head(tickets_len: usize, deltas: &[usize]) -> Vec<usize> {
+    let mut indices_tail = deltas.to_vec();
+    indices_tail.sort_unstable_by(|a, b| b.cmp(a));
+    indices_tail.dedup();
+
+    let mut indices_head: Vec<_> = (0..tickets_len).collect();
```

Comment:
Using swap_remove here may modify the order of the remaining elements, so is this safe? Also, isn't there a better way to do this instead of creating a long vector and then remove elements?